docs/api/Inside-the-Template-Engine.md

```diff
@@ -0,0 +1,340 @@
+[EngineEnvironmentSettings](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Edge/EngineEnvironmentSettings.cs).
+
+It has single constructor that accepts ITemplateEngineHost and set up settings
+based on TemplateEngine.Edge implementation of component manager, paths, and
+environment. All this can also be passed in as optional parameters.
+
+### [ITemplateEngineHost](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Abstractions/ITemplateEngineHost.cs)
+
+The interface is responsible for providing host information to template engine,
+managing file system and do the logging. Host may also provide default parameter
+values for template via TryGetHostParamDefault method.
+
+The applications using template engine are often called as “hosts”.
+dotnet/templating repo is managing one of such hosts: dotnet new CLI, which is
+part of dotnet SDK.
+
+Main host properties:
+
+-   Identifier – unique name of the host
+
+-   Version – version of the host
+
+Those properties are used to identify host to various built-in components
+explained below.
+
+If you are considering using the template engine core, you need to make
+implementation of this interface representing your host. Template engine
+provides default implementation:
+[DefaultTemplateEngineHost](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Edge/DefaultTemplateEngineHost.cs).
+
+#### IPhysicalFileSystem
+
+Abstraction over file system, accessible from ITemplateEngineHost.
+
+Each template engine host should support physical and in-memory file system and
+virtualize file system under given path. Switch between physical and in-memory
+is done via ITemplateEngineHost.VirtualizePath method.
+
+Template engine provides default implementation of file systems:
+[PhysicalFileSystem](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Utils/PhysicalFileSystem.cs)
+and
+[InMemoryFileSystem](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Utils/InMemoryFileSystem.cs).
+
+### [IComponentManager](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Abstractions/IComponentManager.cs)
+
+ComponentManager is responsible for loading components provided by Host from
+ITemplateEngineHost.BuiltIns, but additional components can be added via
+IComponentManager.AddComponent or they are dynamically loaded from
+TemplatePackage during scanning similar to templates. ComponentManager
+implementation is not publicly accessible, however it is used when default
+implementation of IEngineEnvironmentSettings is used.
+
+### [IEnvironment](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Abstractions/IEnvironment.cs)
+
+Abstraction over environment variables. Default implementation uses system
+environment variables.
+
+### [IPathInfo](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Abstractions/IPathInfo.cs)
```

Comment:
Could these be listed at the very end? These look unimportant in understanding the main architecture. They are just simple abstractions the user can figure out while coding. But they make the document hard to follow for one reason: we have too many headers with various depths.

Reply:
They belong to environment settings not sure, if it is a good trade-off